### PR TITLE
55 enforce file extensions

### LINKED ISSUE #55: enforce compatible file extensions for plot backends
Applies to all `RasterPop.plot()` backends:

Feature request to implement a method for ensuring the user passes the correct file extension. In cases where the user tries to do something janky like:

`rp.plot(which=“matplotlib", save=“somewhere/somefile.html")`

_Originally posted by @r-leyshon in https://github.com/datasciencecampus/transport-network-performance/pull/44#discussion_r1288587134_

### PR DESCRIPTION
**Pull request by cbrowning04**
## Description

This issue aims to enforce file extensions in the `.plot()` backends in the `RasterPop` class.  
In order to achieve this, a new defensive function was created to be used throughout the package. This can be found in `defence.py`

Fixes #55 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Test configuration details:
* OS: Windows 10
* Python version: 3.9.13
* Java version: N/A
* Python management system: Conda

## Checklist:

- [x] My code follows the intended structure of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/transport_performance/population/rasterpop.py [hunk 1]
```diff
@@ -738,6 +763,18 @@ def _plot_matplotlib(
         # write to file if filepath is given
         if save is not None:
             _check_parent_dir_exists(save, "save", create=True)
+            # get extension for error messge
+            root, ext = os.path.splitext(save)
+            save = _enforce_file_extension(
+                save,
+                ".png",
```
Comment (SergioRec) [inline src/transport_performance/population/rasterpop.py]: Matplotlib allows saving figures in other formats in addition to `.png`. A list of allowed formats for a specific figure can be retrieved running `plt.gcf().canvas.get_supported_filetypes()`. In the case of `_plot_matplotlib` function, the output of this is:

```
{'eps': 'Encapsulated Postscript', 'jpg': 'Joint Photographic Experts Group', 'jpeg': 'Joint Photographic Experts Group', 'pdf': 'Portable Document Format', 'pgf': 'PGF code for LaTeX', 'png': 'Portable Network Graphics', 'ps': 'Postscript', 'raw': 'Raw RGBA bitmap', 'rgba': 'Raw RGBA bitmap', 'svg': 'Scalable Vector Graphics', 'svgz': 'Scalable Vector Graphics', 'tif': 'Tagged Image File Format', 'tiff': 'Tagged Image File Format', 'webp': 'WebP Image Format'}
```
 Would it be possible to populate the `exp_ext` with the keys of this dictionary to let users save in allowed formats? Happy with `.png` being the default. Something like:
 
```
    def _plot_matplotlib(
        self, save: str = None, figsize: tuple = (6.4, 4.8)
    ) -> Union[plt.Axes, None]:

        # handle kwarg type checks
        _type_defence(figsize, "figsize", tuple)

        # handle matplotlib and rioxarry steps
        fig, ax = plt.subplots(figsize=figsize)
        self._xds.plot(ax=ax)
        plt.tight_layout()

        formats = [
            "." + n for n in
            list(plt.gcf().canvas.get_supported_filetypes().keys())
        ]

        # write to file if filepath is given
        if save is not None:
            _check_parent_dir_exists(save, "save", create=True)
            # get extension for error messge
            root, ext = os.path.splitext(save)
            save = _enforce_file_extension(
                save,
                formats,
                ".png",
                "save",
                msg=(
                    f"Format '{ext}' not implemented. Writing matplotlib plot"
                    " as '.png'"
                ),
            )
            fig.savefig(save)
            ax = None

        return ax
```
(There's probably a better solution for the "." in the extension list...).

File: src/transport_performance/population/rasterpop.py [hunk 2]
```diff
@@ -696,6 +709,18 @@ def _plot_cartopy(
         # get the current figure and resize it to match the axis before saving
         if save is not None:
             _check_parent_dir_exists(save, "save", create=True)
+            # get extension for error messge
+            root, ext = os.path.splitext(save)
+            save = _enforce_file_extension(
+                save,
+                ".png",
```
Comment (SergioRec) [inline src/transport_performance/population/rasterpop.py]: See comment for `matplotlib` charts. Same code works for `cartopy`.

**Review by SergioRec — COMMENTED** [review]
Code and tests work fine, thanks @CBROWN-ONS. 

I've suggested a change to `rasterpop.py` to account for different allowed extensions, as saved figures are now restricted to `.png` format. It may affect tests for the `rasterpop.py` module so not sure if it would be within the scope of this PR. Perhaps @ethan-moss has any suggestion?

File: src/transport_performance/population/rasterpop.py (hunk 1 above)
Comment (cbrowning04) [reply src/transport_performance/population/rasterpop.py]: HI @SergioRec. Originally when discussing with @ethan-moss, he requested to only allow for one file extension (png for cartopy and matplotlib). You may want to discuss with him if you feel that we need to allow more accepted types. I'm happy with both options myself 😄

**Review by SergioRec — APPROVED** [review]
Added support for other file formats on save in matplotlib and cartopy.

**Pull request merged**